src/ir/proto/sql/decode_test.cc

```diff
@@ -0,0 +1,152 @@
+//----------------------------------------------------------------------------
+
+#include "src/ir/proto/sql/decode.h"
+
+#include "google/protobuf/text_format.h"
+#include "google/protobuf/util/message_differencer.h"
+#include "absl/strings/string_view.h"
+#include "src/common/testing/gtest.h"
+#include "src/ir/proto/sql/sql_ir.pb.h"
+
+namespace raksha::ir::proto::sql {
+
+class ComparableSourceTableColumn {
+ public:
+  ComparableSourceTableColumn() = default;
+  ComparableSourceTableColumn(const SourceTableColumn &source_table_column)
+    : column_(source_table_column.column_path()),
+      confidentiality_tags_(source_table_column.confidentiality_tags().begin(),
+                            source_table_column.confidentiality_tags().end()),
+      integrity_tags_(source_table_column.integrity_tags().begin(),
+                      source_table_column.integrity_tags().end())
+      {}
+
+  void AddConfidentialityTag(absl::string_view tag) {
+    confidentiality_tags_.insert(std::string(tag));
+  }
+
+  void AddIntegrityTag(absl::string_view tag) {
+    integrity_tags_.insert(std::string(tag));
+  }
+
+  void SetColumn(absl::string_view column) {
+    CHECK(!column_) << "Attempt to set a column that was already set!";
+    column_ = column;
+  }
+
+  void ExpectEq(const ComparableSourceTableColumn &other) const {
+    EXPECT_EQ(column_, other.column_);
+    EXPECT_EQ(confidentiality_tags_, other.confidentiality_tags_);
+    EXPECT_EQ(integrity_tags_, other.integrity_tags_);
+  }
+
+ private:
+  std::optional<std::string> column_;
+  absl::flat_hash_set<std::string> confidentiality_tags_;
+  absl::flat_hash_set<std::string> integrity_tags_;
+};
+
+// Why are we creating this recursive traversal through the values rather
+// than a visitor? The visitor moves through the code of the IR, visiting
+// Operations, the Modules implementing those Operations, etc. But the
+// properties that we want to verify about the nesting of the claims around a
+// column storage are properties of the Values. In general, traversing values
+// can get quite complicated, as different outputs of an Operation can be
+// sent to different locations and can form loops. Visiting values like this,
+// in fact, starts to look like a dataflow analysis, which we want Souffle to
+// do. Rather than wrestling with the whole problem now, we take advantage of
+// the fact that we expect a single, linear nesting of values which can easily
+// be traversed right here.
+//
+// In addition, the process of traversing the value tree largely overlaps
+// with checking the properties that we want to check anyway.
+//
+// NOTE: It would be nicer to return the result rather than use an output
+// parameter, but that would preclude using the `ASSERT` macros (which expect
+// to be in a function returning void).
+void GatherSourceColumnInfoFromValues(
+    const Value val, ComparableSourceTableColumn &result) {
+  if (const value::StoredValue *storage_val = val.If<value::StoredValue>()) {
```

Comment:
A visitor is probably better than `if..else if.. else`? You could also consider `utils::overloaded`?

Reply:
OK, I'll check it out.

Reply:
Implemented this using a visitor using `utils::overloaded`.